### ISSUE TITLE
TypeError: expected.toLowerCase() is not a function when using "rest.all()"

### ISSUE THREAD

**Issue by rmachado-studocu** [kettanaito/msw#1020]
## Describe the bug
When using `rest.all()` to mock all requests of a given path, it blocks HMR (and other requests).

That's due to an issue with the usage of `isStringEqual` in the file [`src/utils/request/onUnhandledRequest.ts`](https://github.com/mswjs/msw/blob/master/src/utils/request/onUnhandledRequest.ts#L60). 
It assumes that `method` is always a string and with the introduction of `rest.all` it seems to be a `RegExp` now.

## Environment

- `msw: 0.36.0`
- `nodejs: 14.17.6`
- `npm: 6.14.15`
- `yarn: 1.22.15`
- `Chrome: Version 96.0.4664.55 (Official Build) (x86_64)`

## To Reproduce
1. Scaffold a create-react-app
2. Set up MSW with a `rest.all()` handler (mine was like this: `rest.all('/todos', (req, res, ctx) => {...});`
3. Run `yarn start`
4. Make a change on one of the files
5. See error in the console like the ones below:

## Expected behavior
Requests not targeted by the `rest.all()` should still be performed.
By temporarily patching the rest-deps changing the return to:
```js
return actual.toString().toLowerCase() === expected.toString().toLowerCase();
```
solved the issue, but I know that's not the right fix 😄 

## Screenshots
HMR broken:
![image](https://user-images.githubusercontent.com/89906313/145074486-4f9577c4-b048-4fc7-80b0-70f4c10d7135.png)

(also messes requests for the chrome extension)
![image](https://user-images.githubusercontent.com/89906313/145074304-14b938f5-3255-4742-806d-633e83b5c7b2.png)

**Comment by kettanaito** [comment]
Hey, @rmachado-studocu. Thanks for reporting this.

Here's where we're handling the request method matching in the `RestHandler`:

https://github.com/mswjs/msw/blob/10dbde6dc5f447dc1728004b2ca70e1ee2edbed1/src/handlers/RestHandler.ts#L171-L174

I wonder, how come `request.method` becomes a regular expression? That sounds like a no-op to me. I'll follow the steps to reproduce what you've mentioned to see what's going on.

**Comment by rmachado-studocu** [comment]
Hi @kettanaito,

From what I can see, you're creating `all` as RegExp handler: https://github.com/mswjs/msw/blob/master/src/rest.ts#L35

I was checking on the integration tests and it hasn't been caught on them because you always have a fallback - `rest.all('*', () => { ... });` - and you check against that generic response.

The issue here is exactly the case where you set up something like:

```js
rest.all('/specific/endpoint', () => { /*...*/ });
```

And a request is done to say `/another/different/endpoint`... 
Not having any generic request should let it pass and request the actual resource... right?

**Comment by kettanaito** [comment]
Thanks for more insights, @rmachado-studocu.

## Root cause

Yeah, the issue is with the `rest.all`'s method being a regular expression. We do handle regular expressions in `RestHandler` but we _don't_ handle them in `onUnhandledRequest`, which is precisely where the exception originates from:

https://github.com/mswjs/msw/blob/10dbde6dc5f447dc1728004b2ca70e1ee2edbed1/src/utils/request/onUnhandledRequest.ts#L60

> This function processes the handlers, but handlers may contain `rest.all`, where `handler.method` will be a RegExp. This breaks the `isStringEqual` usage here. 

I'm addressing this in #1021. I'm also improving the typings of the said method, and it's already helped me to discover that we have a similar issue with the GraphQL handlers like `graphql.operation` where the operation name is a regular expression.

**Comment by kettanaito** [comment]
The fix has been released in [0.36.2](https://github.com/mswjs/msw/releases/tag/v0.36.2).

**Comment by rmachado-studocu** [comment]
Super! Thanks once again! 😉
